crates/ra_hir/src/expr/validation.rs

```diff
@@ -0,0 +1,67 @@
+use rustc_hash::FxHashSet;
+use crate::{
+    expr::AstPtr,
+    HirDatabase,
+    Function,
+    Name,
+    diagnostics::{DiagnosticSink, MissingFields},
+    adt::AdtDef,
+    Path
+};
+
+use ra_syntax::ast::{AstNode, NamedFieldList};
+
+use super::{Expr, StructLitField, ExprId};
+
+pub(crate) fn validate_body(func: &Function, db: &impl HirDatabase, sink: &mut DiagnosticSink) {
+    let body = func.body(db);
+    for e in body.exprs() {
+        match e {
+            (id, Expr::StructLit { path, fields, spread }) => {
+                validate_struct_literal(id, path, fields, spread, func, db, sink)
+            }
+            _ => (),
+        }
+    }
+}
+
+fn validate_struct_literal(
+    id: ExprId,
+    _path: &Option<Path>,
+    fields: &Vec<StructLitField>,
+    spread: &Option<ExprId>,
+    func: &Function,
+    db: &impl HirDatabase,
+    sink: &mut DiagnosticSink,
```

Comment:
So, one way to redunce a number of arguments here is to define an 

```
struct ExprValidator {
    db: DB,
    func: Function,
    infer: InferenceResult,
    sink: &mut DiagnosticSink
}
```

and make `validate_struct_literal` a method on the struct.

I think it makes sense to do that, I expect we'd want to write quite a few validations. Additionally, the Validator caches queries like `infer`, which also seems like a little, but pleasant win

Reply:
Ugh. That can be a problem for me. I'm quite a newbie in rust and can't figure out how to hold ` &impl HirDatabase` as a field. Compiler doesn't eat it for an obvious reason. So I guess I still need to pass db as an argument. 
The second one: I have to specify DiagnosticSink lifetime explicitly when set it as a field. Like this one
```rust
pub(crate) struct ExprValidator<'a> {
    func: Function,
    infer: Arc<InferenceResult>,
    sink: &'a mut DiagnosticSink<'a>,
}
```
and compiler forces me to specify this lifetime for all diagnostic call stack. This is becoming a mess and I guess I'm doing something wrong.

Reply:
I'm pretty proud of my self. I've changed struct definition to this one
```rust
pub(crate) struct ExprValidator<'a, 'b: 'a> {
    func: Function,
    infer: Arc<InferenceResult>,
    sink: &'a mut DiagnosticSink<'b>,
}
```
and this made the trick

Reply:
Ouch, yeah, nested lifetimes are tricky